src/interface/control_textbox.cpp

```diff
@@ -195,75 +266,151 @@ ControlTextbox::on_key_down(const SDL_KeyboardEvent& key)
   if (!m_has_focus)
     return false;
 
-  if (key.keysym.sym == SDLK_LEFT && m_caret_pos > 0)
+  if (key.keysym.sym == SDLK_LEFT)
   {
-    if (!m_shift_pressed && m_secondary_caret_pos != m_caret_pos)
+    if (!m_shift_pressed && m_secondary_caret.pos != m_caret.pos)
     {
-      m_secondary_caret_pos = m_caret_pos = std::min(m_secondary_caret_pos, m_caret_pos);
+      set_caret_pos(std::min(m_secondary_caret.pos, m_caret.pos));
+      set_secondary_caret_pos(m_caret.pos);
     }
     else
     {
-      m_caret_pos--;
+      set_caret_pos(m_caret.pos - 1);
       m_cursor_timer = CONTROL_CURSOR_TIMER;
       if (!m_shift_pressed)
-        m_secondary_caret_pos = m_caret_pos;
+        set_secondary_caret_pos(m_caret.pos);
     }
 
-    recenter_offset();
+    recenter_offset(true);
     return true;
   }
-  else if (key.keysym.sym == SDLK_RIGHT && m_caret_pos < int(m_charlist.size()))
+  else if (key.keysym.sym == SDLK_RIGHT)
   {
-    if (!m_shift_pressed && m_secondary_caret_pos != m_caret_pos)
+    if (!m_shift_pressed && m_secondary_caret.pos != m_caret.pos)
     {
-      m_secondary_caret_pos = m_caret_pos = std::max(m_secondary_caret_pos, m_caret_pos);
+      set_caret_pos(std::max(m_secondary_caret.pos, m_caret.pos));
+      set_secondary_caret_pos(m_caret.pos);
     }
     else
     {
-      m_caret_pos++;
+      set_caret_pos(m_caret.pos + 1);
       m_cursor_timer = CONTROL_CURSOR_TIMER;
       if (!m_shift_pressed)
-        m_secondary_caret_pos = m_caret_pos;
+        set_secondary_caret_pos(m_caret.pos);
     }
 
-    recenter_offset();
+    recenter_offset(true);
     return true;
   }
-  else if (key.keysym.sym == SDLK_BACKSPACE)
+  else if (key.keysym.sym == SDLK_UP)
```

Comment:
Why don't we even use switch/case in the code for these, just food for thought? A compiler probably optimizes it to such anyway, hopefully...